### ISSUE TITLE
Contributing SigLIP to `mlx-examples`

### ISSUE THREAD

**Issue by suvadityamuk** [ml-explore/mlx-examples#747]
Hello!

I'm new to the MLX ecosystem, and I came across the fact that there is a valid CLIP implementation available in the repository. Keeping in mind that [SigLIP](https://arxiv.org/abs/2303.15343) has the same structure as that of CLIP ([as per the HF implementation available](https://github.com/huggingface/transformers/blob/v4.40.1/src/transformers/models/siglip/modeling_siglip.py)), I was wondering if there would be any interest for converting the SigLIP weights to a MLX-compatible format? 

Many community modelling efforts to create MLLMs have been consistently choosing SigLIP over the original CLIP model due to better performance.

Happy to discuss and understand if there are any nuances I may have overlooked, or should keep in mind. Thank you so much!
(and thank you so much for MLX, such a life-saver for folks doing ML dev on Apple Silicon!)

Thoughts? @awni

**Comment by awni** [comment]
Definitely! If they run out of the box with the MLX Clip example from Hugging Face that would be awesome.

You can upload them to the HF MLX Community (https://huggingface.co/mlx-community/) or to your own HF space. Either is ok!

**Comment by x4080** [comment]
@suvadityamuk how do you convert siglip to mlx ? I tried using 
python convert.py --hf-repo google/siglip-base-patch16-224
and not success, merges.txt not found or something

**Comment by sujantkumarkv** [comment]
i was thinking along the exact same lines as @suvadityamuk. what's the update here or if i shall work on it?

**Comment by awni** [comment]
No update that I know of. I think if its a small modification of the Clip example we could include it there?

**Comment by suvadityamuk** [comment]
@awni @sujantkumarkv Been trying to use it out-of-the-box but was facing those issues. Have been caught up for a bit trying to experiment but no luck. It seems there would have to be some modifications done to the existing CLIP example to get it to work.

**Comment by sujantkumarkv** [comment]
_I'm asking to see if i'm in the right direction here. Feel free to teach me here, not an expert._

> No update that I know of. I think if its a small modification of the Clip example we could include it there?

based on my reading, SigLIP paper only is a small tweak which uses sigmoid loss against the CLIP's softmax which depended on a global state. Thus, it helps in handling much larger batch sizes but also maintaining the performance on smaller batches during *pretraining*. So, the sigLip example might include the changes loss function, probably [here](https://github.com/ml-explore/mlx-examples/blob/main/clip/model.py#L131) and maybe some more slight changes (need to read paper in depth)

> Been trying to use it out-of-the-box but was facing those issues. Have been caught up for a bit trying to experiment but no luck. It seems there would have to be some modifications done to the existing CLIP example to get it to work.

Can you elaborate on what exactly you tried and what didn't work? I didn't read the transformer's implementation full & maybe I'm wrong but CLIP & SigLIP must be almost similar with little changes similar to what I said?

cc @awni @suvadityamuk

**Comment by awni** [comment]
I haven't read the paper so I can't confirm. But if it indeed is a small delta from CLIP then it could make sense to support it in the CLIP example. If not, its probably not worth including in MLX examples. It would be great to have something like a mlx-clip package maintained by the community which can load all of those clip models!

**Comment by sujantkumarkv** [comment]
okay will look into this.

**Comment by aavetis** [comment]
thoughts on adding Apple's proper [mobileclip](https://github.com/apple/ml-mobileclip) instead? 

> Our smallest variant MobileCLIP-S0 obtains similar zero-shot performance as [OpenAI](https://arxiv.org/abs/2103.00020)'s ViT-B/16 model while being 4.8x faster and 2.8x smaller.
MobileCLIP-S2 obtains better avg zero-shot performance than [SigLIP](https://arxiv.org/abs/2303.15343)'s ViT-B/16 model while being 2.3x faster and 2.1x smaller, and trained with 3x less seen samples.
MobileCLIP-B(LT) attains zero-shot ImageNet performance of 77.2% which is significantly better than recent works like [DFN](https://arxiv.org/abs/2309.17425) and [SigLIP](https://arxiv.org/abs/2303.15343) with similar architectures or even [OpenAI's ViT-L/14@336](https://arxiv.org/abs/2103.00020).

**Comment by maxlund** [comment]
Any updates on this? Would love to run other CLIP variants via MLX, any advice would be appreciated!

**Comment by maxlund** [comment]
I think this PR might shed some light on what is required to convert a SigLIP model to MLX, since @Blaizzy did it as part of adding PaliGemma to the mlx-vlm repo (in which SigLIP is a component). See this PR https://github.com/Blaizzy/mlx-vlm/pull/24
